### PR TITLE
feat(pgwire): replace BoxStream in PgResult with static type

### LINKED ISSUE #5624: frontend: support concrete stream type rather than using BoxStream
### Is your feature request related to a problem? Please describe.

as discuss in #5556 
We use BoxStream to transfer data now. But Boxstream will impact the performance because it's a dynamic bind. Boxstream is Box, it's a dynamic bind. So I guess every time we call 'stream.next()' to get a row, it needs to call by some mechanism like vtable. Hence if there are millions of rows in the result set, we may need to access vtable millions of times.

### Describe the solution you'd like

``The solution is to implement the concrete type which implements stream trait, such as:
```
struct LocalQueryStream {
 ...
}
impl Stream for LocalQueryStream{
  fn poll_next(){
   ...
  }
}
```
so that we can use the concrete result type instead of BoxStream to prevent dynamic binding. 
```
// original version
struct PgResponse{
  ...
  result: BoxStream<>
}

// new version 
struct PgResponse{
  result: LocalQueryStream
}
```

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Gun9niR**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Previously pgwire uses `BoxStream` as `value_stream` type. If we have lots of rows, dynamic dispatch may bring some overhead. Therefore, we change `BoxStream` to a generic param, and use enum dispatch in frontend.

### Performance

Because this PR involves interface changes, benchmark code is not included. The micro bench are as follows.

- This PR

```rust
// One million chunks.
const CHUNK_NUM: usize = 1000000;

fn get_chunk() -> DataChunk {
    // Make the chunk as small as possible because we are more interested in the overhead of polling
    // the stream.
    DataChunk::from_rows(&[risingwave_common::array::Row::new(vec![])], &[])
}

/// Format scalars according to postgres convention.
fn pg_value_format(d: ScalarRefImpl<'_>, format: bool) -> Bytes {
    // format == false means TEXT format
    // format == true means BINARY format
    if !format {
        match d {
            ScalarRefImpl::Bool(b) => if b { "t" } else { "f" }.into(),
            _ => d.to_string().into(),
        }
    } else {
        d.binary_serialize()
    }
}

pub fn to_pg_rows(chunk: DataChunk, format: bool) -> Vec<Row> {
    chunk
        .rows()
        .map(|r| {
            Row::new(
                r.values()
                    .map(|data| data.map(|data| pg_value_format(data, format)))
                    .collect_vec(),
            )
        })
        .collect_vec()
}

#[try_stream(ok = DataChunk, error = RwError)]
async fn yield_chunks(chunks: Vec<DataChunk>) {
    for chunk in chunks {
        yield chunk;
    }
}

fn bench_local(c: &mut Criterion) {
    let runtime = tokio::runtime::Builder::new_current_thread()
        .enable_all()
        .build()
        .unwrap();

    c.bench_function("bench_local", |b| {
        b.to_async(&runtime).iter_custom(|iters| async move {
            let mut total_time = Duration::default();
            for _ in 0..iters {
                let mut chunks = vec![];
                for _ in 0..CHUNK_NUM {
                    chunks.push(get_chunk());
                }
                let mut query_stream = PgResponseStream::LocalQuery(LocalQueryStream::new(
                    Box::pin(yield_chunks(chunks)),
                    true,
                ));

                let start = Instant::now();
                while let Some(_) = query_stream.next().await {}
                total_time += start.elapsed();
            }
            total_time
        })
    });
}

fn bench_distributed(c: &mut Criterion) {
    let runtime = tokio::runtime::Builder::new_current_thread()
        .enable_all()
        .build()
        .unwrap();

    c.bench_function("bench_distributed", |b| {
        b.to_async(&runtime).iter_custom(|iters| async move {
            let mut total_time = Duration::default();
            for _ in 0..iters {
                let (chunk_tx, chunk_rx) =
                    tokio::sync::mpsc::channel::<SchedulerResult<DataChunk>>(CHUNK_NUM);
                for _ in 0..CHUNK_NUM {
                    chunk_tx.send(Ok(get_chunk())).await.unwrap();
                }
                drop(chunk_tx);
                let mut query_stream =
                    PgResponseStream::DistributedQuery(DistributedQueryStream::new(chunk_rx, true));

                let start = Instant::now();
                while let Some(_) = query_stream.next().await {}
                total_time += start.elapsed();
            }
            total_time
        })
    });
}

fn bench_query_stream(c: &mut Criterion) {
    bench_local(c);
    bench_distributed(c);
}

criterion_group!(benches, bench_query_stream);
criterion_main!(benches);

```

- main

```rust
// One million chunks.
const CHUNK_NUM: usize = 1000000;

fn get_chunk() -> DataChunk {
    // Make the chunk as small as possible because we are more interested in the overhead of polling
    // the stream.
    DataChunk::from_rows(&[risingwave_common::array::Row::new(vec![])], &[])
}

/// Format scalars according to postgres convention.
fn pg_value_format(d: ScalarRefImpl<'_>, format: bool) -> Bytes {
    // format == false means TEXT format
    // format == true means BINARY format
    if !format {
        match d {
            ScalarRefImpl::Bool(b) => if b { "t" } else { "f" }.into(),
            _ => d.to_string().into(),
        }
    } else {
        d.binary_serialize()
    }
}

pub fn to_pg_rows(chunk: DataChunk, format: bool) -> Vec<Row> {
    chunk
        .rows()
        .map(|r| {
            Row::new(
                r.values()
                    .map(|data| data.map(|data| pg_value_format(data, format)))
                    .collect_vec(),
            )
        })
        .collect_vec()
}

#[try_stream(ok = DataChunk, error = RwError)]
async fn yield_chunks(chunks: Vec<DataChunk>) {
    for chunk in chunks {
        yield chunk;
    }
}

#[try_stream(ok = Vec<Row>, error = BoxedError)]
async fn stream_row_inner(data_stream: BoxedDataChunkStream, format: bool) {
    #[for_await]
    for chunk in data_stream {
        let rows = to_pg_rows(chunk?, format);
        yield rows;
    }
}

fn bench_local(c: &mut Criterion) {
    let runtime = tokio::runtime::Builder::new_current_thread()
        .enable_all()
        .build()
        .unwrap();

    c.bench_function("bench_local", |b| {
        b.to_async(&runtime).iter_custom(|iters| async move {
            let mut total_time = Duration::default();
            for _ in 0..iters {
                let mut chunks = vec![];
                for _ in 0..CHUNK_NUM {
                    chunks.push(get_chunk());
                }
                let mut query_stream =
                    Box::pin(stream_row_inner(Box::pin(yield_chunks(chunks)), true));

                let start = Instant::now();
                while let Some(_) = query_stream.next().await {}
                total_time += start.elapsed();
            }
            total_time
        })
    });
}

#[try_stream(ok = Vec<Row>, error = BoxedError)]
async fn stream_from_channel_inner(
    mut chunk_rx: tokio::sync::mpsc::Receiver<SchedulerResult<DataChunk>>,
    format: bool,
) {
    while let Some(chunk_inner) = chunk_rx.recv().await {
        let chunk = chunk_inner?;
        let rows = to_pg_rows(chunk, format);
        yield rows;
    }
}

fn bench_distributed(c: &mut Criterion) {
    let runtime = tokio::runtime::Builder::new_current_thread()
        .enable_all()
        .build()
        .unwrap();

    c.bench_function("bench_distributed", |b| {
        b.to_async(&runtime).iter_custom(|iters| async move {
            let mut total_time = Duration::default();
            for _ in 0..iters {
                let (chunk_tx, chunk_rx) =
                    tokio::sync::mpsc::channel::<SchedulerResult<DataChunk>>(CHUNK_NUM);
                for _ in 0..CHUNK_NUM {
                    chunk_tx.send(Ok(get_chunk())).await.unwrap();
                }
                drop(chunk_tx);
                let mut query_stream = Box::pin(stream_from_channel_inner(chunk_rx, true));

                let start = Instant::now();
                while let Some(_) = query_stream.next().await {}
                total_time += start.elapsed();
            }
            total_time
        })
    });
}

fn bench_query_stream(c: &mut Criterion) {
    bench_local(c);
    bench_distributed(c);
}

criterion_group!(benches, bench_query_stream);
criterion_main!(benches);
```

- bench result

```
Benchmarking bench_local: Collecting 100 samples in estimate                                                            bench_local             time:   [67.611 ms 67.632 ms 67.655 ms]
                        change: [-2.6439% -2.5930% -2.5418%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  4 (4.00%) high mild
  1 (1.00%) high severe

Benchmarking bench_distributed: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 34.8s, or reduce sample count to 10.
Benchmarking bench_distributed: Collecting 100 samples in es                                                            bench_distributed       time:   [116.41 ms 116.47 ms 116.52 ms]
                        change: [-21.287% -21.214% -21.134%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild
```

## Checklist

- [x] I have written necessary rustdoc comments
- ~[ ] I have added necessary unit tests and integration tests~
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

## Refer to a related PR or issue link (optional)

Closes #5624

### CODE REVIEW

**Comment by Gun9niR** [comment]
aa

**Review by ZENOTME — APPROVED** [review]
Cool!LGTM @BowenXiao1999 PTAL

File: src/frontend/src/handler/mod.rs [hunk 1]
```diff
@@ -49,12 +57,40 @@ mod show;
...
+    LocalQuery(LocalQueryStream),
+    DistributedQuery(DistributedQueryStream),
+    Rows(BoxStream<'static, RowSetResult>),
+}
+
+impl Stream for PgResponseStream {
+    type Item = std::result::Result<Vec<Row>, BoxedError>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        match self.project() {
```
Comment (StrikeW) [inline src/frontend/src/handler/mod.rs:74]: This is interesting. So without the pin_project crate we cannot do the matching here? What is the reason behind the scene?🤔

Comment (Gun9niR) [reply src/frontend/src/handler/mod.rs]: It turns out matching can be done without pin_project. I had a wrong understanding of `Pin` previously.

File: src/utils/pgwire/src/pg_server.rs [hunk 2]
```diff
@@ -143,12 +155,13 @@ mod tests {
...
 
     #[async_trait::async_trait]
-    impl Session for MockSession {
+    impl Session<BoxStream<'static, RowSetResult>> for MockSession {
         async fn run_statement(
             self: Arc<Self>,
             sql: &str,
             _format: bool,
-        ) -> Result<PgResponse, Box<dyn Error + Send + Sync>> {
+        ) -> Result<PgResponse<BoxStream<'static, RowSetResult>>, Box<dyn Error + Send + Sync>>
```
Comment (BowenXiao1999) [inline src/utils/pgwire/src/pg_server.rs:163]: Can we also get rid of the BoxStream here?

**Review by BowenXiao1999 — APPROVED** [review]
Maybe we can consider add micro bench just like batch executors later. LGTM!

**Comment by Gun9niR** [comment]
> Maybe we can consider add micro bench just like batch executors later. LGTM!

The difficulty is that optimizations like this involve interface changes, so it's very hard to make the bench code stable.

File: src/utils/pgwire/src/pg_server.rs (hunk 2 above)
Comment (Gun9niR) [reply src/utils/pgwire/src/pg_server.rs:163]: Is performance important for mocked session 🤣 ?

Comment (BowenXiao1999) [reply src/utils/pgwire/src/pg_server.rs:163]: Oops, my bad 🤣

**Pull request merged**